### ISSUE TITLE
Fine-tune yields worse COCO AR and AP

### ISSUE THREAD

**Issue by briliantnugraha** [niki-amini-naieni/CountGD#22]
Dear @niki-amini-naieni , thanks for the paper, and it's been a while since my last consults with you regarding CountGD paper in [HF](https://huggingface.co/spaces/nikigoli/countgd/discussions/2#671a155c447c240da4255eb5).

Sorry to bother you again, but I encounter an issue, where I am trying to finetune with both pretrained model (**A: checkpoint_fsc147_best**) or default model (**B: groundingdino_swinb_cogcoor**), but I get worse results with default train.sh configuration.

FYI, the dataset that I am using for training has 6-8K images, and there are 5-100 objects in one image.
And for training, I follow your train.sh guidance by using **odvg** format, followed by validation with **coco** format.

However, the end results (both tested with **A** or **B** model is a good first two epochs (high AP and AR). Followed by steadily worse AP and AR subsequently.

Do you have any tips on what things that I might omit? or do I need to adjust the some hyperparameters that is more suitable for my dataset?

Thanks for your help, and looking forward for your reply.
Best regards,
Brilian

**Comment by niki-amini-naieni** [comment]
Hi, could you please let me know how the counting error (MAE and RMSE) change for each epoch? Could you attach the training log? Also, are you using visual exemplars or text only or both? The training code expects both.

**Comment by briliantnugraha** [comment]
Hi @niki-amini-naieni , below is the training log.
Hope it helps, let me know if you need further info, thanks.
[training.log](https://github.com/user-attachments/files/18204704/training.log)

**Comment by niki-amini-naieni** [comment]
The training code regresses the **center points** of the bounding boxes, not the full bounding boxes. This is because the FSC-147 training set does not provide the full bounding box annotations (it only provides the center points of the objects). This is what could be causing the AP and AR to drop. The counting error is what CountGD is trained and expected to reduce. I also did not check/verify the AP and AR calculations for each epoch since I did not attempt to obtain bounding boxes from CountGD. These calculations are left over from the Open GroundingDINO codebase. If your dataset provides the bounding boxes, you can modify the training code to regress those as well to improve the AP/AR calculations. You can do this by modifying this line in the training code: 
<img width="542" alt="image" src="https://github.com/user-attachments/assets/d3a0e6cb-9d44-4b3e-a278-9f74dd1651e1" />

**Comment by briliantnugraha** [comment]
Cool, so I just need to replace **[:, :2]** to **[:, :4]** right?
I've noticed that the target_boxes array Nx4 with [cx, cy, w, h] format. So I "guess" that extending it to all of the array should do the job, but cmiiw :)

**Comment by niki-amini-naieni** [comment]
Hopefully, this fixes your issue. Just make sure that the bounding box format is compatible with the one used in the training code. I believe we follow the ODVG format described here: https://github.com/longzw1997/Open-GroundingDino/blob/main/data_format.md#odvg-dataset-format

**Comment by briliantnugraha** [comment]
Hi @niki-amini-naieni, I've tested it, and I confirm that the COCO AP and AR do increase overtime with default training setting that you shared in the repo, thanks again for the guidance !

NOTE:
The only weird thing that I finally find out (which is why it's taken me so long to reply), is that when I use batch size 8, COCO AP and AR will go to 0.0. And I increase this batch size due to the **hope** of getting **faster** training time, which ends up backfiring me last week @@.

My guess is that it **might** be due to gradient exploding issue. Or maybe, have you tested with bigger batch sizes without any issue @niki-amini-naieni ?

**Comment by niki-amini-naieni** [comment]
Hi @briliantnugraha, thanks for the update! I have not tried larger batch sizes. This sounds like something I should look into in the future (:

**Comment by niki-amini-naieni** [comment]
Hi @briliantnugraha, I wonder if the AP calculation at the end of each epoch is accurate since it assumes a "maxDets" of 100 objects, while CountGD (with the default parameters) outputs at most 900 detections per image. One way to fix this would be to change the "num_select" parameter in the CountGD config file to 100 to limit the number of detections output by CountGD. Here is an example of where you could modify "num_select": https://github.com/niki-amini-naieni/CountGD/blob/9d7b75c0f53d8be3e3f6a04428037d4587b52f5b/config/cfg_fsc147_vit_b.py#L88.

**Comment by briliantnugraha** [comment]
Hi @niki-amini-naieni , thanks for the tips.
Luckily, I tend to prioritize AR more, and ignore AP for now, because AR/Average Recall shows the upper bound of the model capability to detect the object of interest :). cmiiw

**Comment by FightingKai01** [comment]
@briliantnugraha 
Hi, sorry to bother you, I referred to your [discussion](https://github.com/niki-amini-naieni/CountGD/issues/22#issuecomment-2556571813) and it helped me a lot.

[training_14.log](https://github.com/user-attachments/files/19416500/training_14.log) is part of my training log on **custom dataset.**Fine-tuning on object detection task using checkpoint_fsc147_best。I found that the AP and AR in the COCO evaluation standard are almost always the following: Very strange results. The entire bounding box is regressed here.
![Image](https://github.com/user-attachments/assets/d326cd5a-c9ef-4bf1-8bbe-837565a834a8)

Do you know the reason for this?Hope to get your help.

**Comment by FightingKai01** [comment]
Thanks for your response.This problem has troubled me for a long time, but there is still no progress.

I did use the COCO dataset.I have noticed the problem you mentioned, and I have checked it again and it complies with the FSC147 data annotation format.

I also hope that I can fine-tune the model successfully to make it suitable for object detection. How can I do this? What files do I need to provide you to help me find the problem?
